3.1 - Console Helper (Formerly Console macro)

markstory opened the pull request:
Implements most of the changes proposed in #6499. I've not implemented the progress macro, as I wanted to keep this smallish so it would be easier to discuss. I'll implement the progress macro in a separate pull request.

markstory committed:
Start adding MacroRegistry and tests.

The MacroRegistry will allow macros to be wired into ConsoleIo in the
future.

markstory committed:
Add macro integration into ConsoleIo.

markstory committed:
Add TableMacro

Add a pretty table formatting macro.

markstory committed:
Use table macro in RoutesShell.

dereuromark commented on src/Shell/Macro/TableMacro.php:
```diff
@@ -0,0 +1,95 @@
+        $widths = [];
+        foreach ($rows as $line) {
+            for ($i = 0, $len = count($line); $i < $len; $i++) {
+                $columnLength = strlen($line[$i]);
```
Should we use mb for utf8 chars?

markstory replied:
Yes we should.

antograssiot commented on tests/TestCase/Shell/Macro/TableMacroTest.php:
```diff
@@ -0,0 +1,83 @@
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP Project
+ * @since         3.0.0
```
3.1.0

antograssiot commented on src/Shell/Macro/TableMacro.php:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    /**
+     * Output a row separator.
```
Output a row

antograssiot commented on src/Shell/RoutesShell.php:
```diff
@@ -34,12 +34,13 @@ class RoutesShell extends Shell
         }
-
-        $this->_outWithColumns($output);
+        $this->_io->macro('table', $output);
```
@markstory I've tried to replace this line with
```php
$this->_io->table($output);
```
as I understood it was working but it's no as it seems to be an extra array around args. They should be the same right or am I doing something wrong ?

markstory replied:
That should be how it works. I will add more tests.

antograssiot commented on src/Console/ConsoleIo.php:
```diff
@@ -363,4 +373,33 @@ public function setLoggers($enable)
+    }
+
+    /**
+     * Conveinence wrapper around macro()
```
Convenience

markstory committed:
Fix up code review issues.

markstory committed:
Rename Macro -> Helper.

A few people think that Macro is the wrong term to use as it implies
C/C++/Lisp macros which is nothing like this feature. By using Helper,
we can better convey what the classes are for (helping in shells). I've
kept the shortform output() method, and also made it possible to get
a helper instance in case someone wants to get instances and call
non-output() methods on it.

markstory commented:
Updated to use `Helper` instead of `Macro`. I've also fixed the argument handling issue that @antograssiot pointed out.

antograssiot committed:
more macro->helper + CS

lorenzo commented on src/Console/Helper.php:
```diff
@@ -0,0 +1,45 @@
+     *
+     * @param \Cake\Console\ConsoleIo $io An io instance.
+     */
+    public function __construct(ConsoleIo $io)
```
Should it accept options like the rest of the registry objects?

markstory replied:
Yes, that be food for settings/config data.

markstory replied:
Fixed.

dereuromark commented on src/Shell/Helper/TableHelper.php:
```diff
@@ -0,0 +1,100 @@
+    {
+        $out = '';
+        foreach ($row as $i => $column) {
+            $out .= '| ' . str_pad($column, $widths[$i], ' ', STR_PAD_RIGHT) . ' ';
```
According to http://stackoverflow.com/questions/14773072/php-str-pad-unicode-issue and the selected answer we need a mb aware helper function here (stupid PHP5...).

markstory replied:
I was wondering the same thing. It might be better to just skip str_pad and implement this in a unicode compatible way.

markstory replied:
Fixed.

markstory committed:
Add InstanceConfig to ConsoleHelpers.

Implement InstanceConfig makes ConsoleHelpers work the same as many
other objects in CakePHP. This consistency is nice for developers.

markstory committed:
Fix UTF8 issues in TableHelper.

str_pad() counts bytes not visible characters. Also add a test for
output() being called with a wrapping array which happens in
the ConsoleIo::__call() code path.

lorenzo committed:
Removing useless link

lorenzo committed:
Removing another useless link

lorenzo commented:
Great job @markstory :clap:

markstory commented:
I'll get started on the progress helper and docs.
